data_steward/cdr_cleaner/clean_cdr_engine.py

```diff
@@ -37,129 +41,225 @@ def add_console_logging(add_handler):
-                                     destination_table_id=destination_table,
-                                     retry_count=retry,
-                                     write_disposition=disposition,
-                                     destination_dataset_id=destination_dataset,
-                                     batch=batch)
 
-        except (oauth2client.client.HttpAccessTokenRefreshError,
-                googleapiclient.errors.HttpError) as exp:
+def validate_params(clazz, **kwargs):
+    """
+    Filters kwargs based on the signature of the 'clazz'
 
-            LOGGER.exception(
-                format_failure_message(project_id=project,
-                                       statement=statement,
-                                       exception=exp))
-            failures += 1
-            continue
-
-        # wait for job to finish
-        query_job_id = results['jobReference']['jobId']
-        incomplete_jobs = bq_utils.wait_on_jobs([query_job_id])
-        if incomplete_jobs:
-            failures += 1
-            raise bq_utils.BigQueryJobWaitError(incomplete_jobs)
-
-        # check if the job is complete and an error has occurred
-        is_errored, error_message = bq_utils.job_status_errored(query_job_id)
-
-        if is_errored:
-            LOGGER.error(
-                format_failure_message(project_id=project,
-                                       statement=statement,
-                                       exception=error_message))
-            failures += 1
-        else:
-            if destination_table is not None:
-                updated_rows = results.get("totalRows")
-                if updated_rows is not None:
-                    LOGGER.info(
-                        f"Query returned {updated_rows} rows for {destination_dataset}.{destination_table}"
-                    )
-
-            successes += 1
-
-    if successes > 0:
-        LOGGER.info(
-            f"Successfully applied {successes} clean rules for {project}.{data_stage}"
-        )
-    else:
-        LOGGER.warning(
-            f"No clean rules successfully applied to {project}.{data_stage}")
+    :param clazz: Clean class or clean function to check the signature of
+    :param kwargs: optional arguments provided by the user
+    :return: filtered dictionary of kwargs
+    :raises ValueError: if a required param for 'clazz' is missing from kwargs
+    """
+    params = inspect.signature(clazz).parameters
+    mandatory_params = ['project_id', 'dataset_id']
+    rule_params = [param for param in params if param not in mandatory_params]
+    # filter kwargs based on required params
+    kwargs = {k: v for k, v in kwargs.items() if k in rule_params}
+    if any(param not in kwargs.keys() for param in rule_params):
+        raise ValueError(
+            f'Params {[param for param in rule_params if param not in kwargs.keys()]} '
+            f'not provided for cleaning rule {clazz.__name__}')
```

Comment:
Yes! This should allow inversion of control such that the rules being applied dictate to the runner the required args. Very cool!

Based on the docs I put together this will need to handle parameters with defaults- perhaps something like this:
```suggestion
    rule_params = {k: v for k, v in params.items() if k not in mandatory_params}
    # filter kwargs based on required params
    kwargs = {k: v for k, v in kwargs.items() if k in rule_params.keys()}
    missing = [k for k, v in rule_params.items() 
               if k not in kwargs.keys() and v is inspect.Parameter.empty]
    if any(missing):
        raise ValueError(
            f'Params {missing} '
            f'not provided for cleaning rule {clazz.__name__}')
```

For testability I recommend extracting out a `get_all_params` function:
```python
def get_all_params(clazz) -> List[str]:
"""
Get all parameters required by the specified cleaning rule
"""
``` 
Naming. Maybe call it `get_custom_kwargs` rather than `validate_params` :man_shrugging:? Naming.